File: USACO/USACO 2020 US Open Bronze Social Distancing I.py
import sys, collections

sys.stdin = open('socdist1.in', 'r')
sys.stdout = open('socdist1.out', 'w')

input = sys.stdin.readline

n = int(input())
stalls = [True if i == '1' else False for i in input().strip()]

gaps = []
prev = 0
for i in range(n):
    if stalls[i]:
        gaps.append(i-prev)
        prev = i
gaps.append(i-prev)
'''
#print(gaps)

ans = 0
if gaps[0] > 0:
    if gaps[-1] > 0:
        ans = max(ans, min(gaps))
    else:
        ans = max(ans, min(gaps[:len(gaps)-1]))
elif gaps[-1] > 0:
    ans = max(ans, min(gaps[1:]))

for i in range(len(gaps)):
    if gaps[i] // 2 == 0: continue
    for j in range(i+1, len(gaps)):
        if gaps[j] // 2 == 0: continue
        ans = max(ans, gaps[i]//2, gaps[j]//2)

print(ans)

'''
def calc_dist():
    min_dist = n+1
    for i in range(n):
        if not stalls[i]:
            continue
        for j in range(i+1, n):
                if not stalls[j]:
                    continue
                min_dist = min(min_dist, j-i)
    return min_dist

ans = 0

if n <= 100:
    for i in range(n):
        if stalls[i]:
            continue
        for j in range(i+1, n):
            if stalls[j]:
                continue
            stalls[i], stalls[j] = True, True
            ans = max(ans, calc_dist())
            stalls[i], stalls[j] = False, False
else:
    if gaps[0] > 0:
        if gaps[-1] > 0:
            ans = max(ans, min(gaps))
        else:
            ans = max(ans, min(gaps[:len(gaps)-1]))
    elif gaps[-1] > 0:
        ans = max(ans, min(gaps[1:]))

    for i in range(len(gaps)):
        if gaps[i] // 2 == 0: continue
        for j in range(i+1, len(gaps)):
            if gaps[j] // 2 == 0: continue
            ans = max(ans, gaps[i]//2, gaps[j]//2)

print(ans)

File: USACO/Bronze/C++/USACO 2020 US Open Bronze Social Distancing I.cpp
#include <bits/stdc++.h>
#include <vector>
using namespace std;

int calc_dist(int n, vector<bool> arr){
	int ret = n+1;
	for (int i = 0; i < n; i++) {
		if (!arr[i]) continue;
		for (int j = i+1; j < n; j++) {
			if (!arr[j]) continue;
			ret = min(ret, j-i);
		}
	}
	return ret;
}

int main() {
	ifstream fin("socdist1.in");
	ofstream fout("socdist1.out");

	int n; fin >> n;
	string s; fin >> s;
	vector<bool> stalls(n, false);
	for (int i = 0; i < n; i++) {
		if (s[i] == '1') stalls[i] = true;
	}

	int ans = 0;
	for (int i = 0; i < n; i++) {
		if (stalls[i]) continue;
		for (int j = i+1; j < n; j++) {
			if (stalls[j]) continue;
			stalls[i] = true;
			stalls[j] = true;
			ans = max(ans, calc_dist(n, stalls));
			stalls[i] = false;
			stalls[j] = false;
		}
	}
	
	fout << ans << '\n';
	
}
